File: Assets/Scripts/Module/Invoke/FActions/TransformToAction.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace FAction
{
    public class TransformToAction : FAction
    {
        public enum TransformControlType
        {
            LocalPosition = 0,
            Position,
            LocalScale,
            LocalRotation,
            Rotation,
        }

        public override ActionType ActionType =>  ActionType.TransformTo;


        public TransformControlType ControlType { get; private set; }

        Transform transform;
        Vector3 start;
        Vector3 target;
        public float time;
       
        float currentTime = 0;

        public void InitTransform(Transform transform, TransformControlType type, Vector3 target, float time = 1)
        {
            this.transform = transform;
            ControlType = type;
            this.target = target;
            this.time = time;
        }

        public override void OnStartTick()
        {
            start = GetTransform();
        }

        public override void Replay()
        {
            base.Replay();
            currentTime = 0;
        }

        public override void Clear()
        {
            base.Clear();
            transform = null;
            start = Vector3.zero;
            time = 0;
        }


        void SetTransform(Vector3 current)
        {
            switch (ControlType)
            {
                case TransformControlType.LocalPosition:
                    transform.localPosition = current;
                    break;
                case TransformControlType.Position:
                    transform.position = current;
                    break;
                case TransformControlType.LocalScale:
                    transform.localScale = current;
                    break;
                case TransformControlType.LocalRotation:
                    transform.localEulerAngles = current;
                    break;
                case TransformControlType.Rotation:
                    transform.eulerAngles = current;
                    break;
                default:
                    break;
            }
        }

        Vector3 GetTransform()
        {
            switch (ControlType)
            {
                case TransformControlType.LocalPosition:
                    return transform.localPosition;
                case TransformControlType.Position:
                    return transform.position;
                case TransformControlType.LocalScale:
                    return transform.localScale;
                case TransformControlType.LocalRotation:
                    return transform.localEulerAngles;
                case TransformControlType.Rotation:
                    return transform.eulerAngles;
                default:
                    break;
            }
            return Vector3.zero;
        }

        public override void Tick()
        {
            base.Tick();
            if (time <= 0)
            {
                SetTransform(target);
                this.IsFinish = true;
                return;
            }
            currentTime += Time.deltaTime;

            float percent = currentTime / time;
            if (percent > 1)
            {
                SetTransform(target);
                this.IsFinish = true;
                return;
            }
          
            var current = GetTransform();
            var lerp = Vector3.Slerp(start, target, percent);
            SetTransform(lerp);

        }
    }
}

File: Assets/.lpproj/Apis/Unity/CS/FAction/TransformToAction.lua
---@class CS.FAction.TransformToAction : CS.FAction.FAction
CS.FAction.TransformToAction = {}

---@field public CS.FAction.TransformToAction.time : CS.System.Single
CS.FAction.TransformToAction.time = nil

---@property readonly CS.FAction.TransformToAction.ActionType : CS.FAction.ActionType
CS.FAction.TransformToAction.ActionType = nil

---@property readwrite CS.FAction.TransformToAction.ControlType : CS.FAction.TransformControlType
CS.FAction.TransformToAction.ControlType = nil

---@return CS.FAction.TransformToAction
function CS.FAction.TransformToAction()
end

---@param transform : CS.UnityEngine.Transform
---@param type : CS.FAction.TransformControlType
---@param target : CS.UnityEngine.Vector3
---@param time : CS.System.Single
function CS.FAction.TransformToAction:InitTransform(transform, type, target, time)
end

function CS.FAction.TransformToAction:OnStartTick()
end

function CS.FAction.TransformToAction:Replay()
end

function CS.FAction.TransformToAction:Clear()
end

function CS.FAction.TransformToAction:Tick()
end